tower-ready-cache/src/service.rs

```diff
@@ -0,0 +1,380 @@
+    ///
+    /// If the service errors, it is removed and dropped and the error is returned.
+    ///
+    /// Otherwise, the `next_ready_index` is set with the provided index and
+    /// `Async::Ready` is returned.
+    pub fn poll_next_ready_index(&mut self, index: usize) -> Poll<(), error::Failed<K>> {
+        self.next_ready_index = None;
+
+        let (_, svc) = self.ready.get_index_mut(index).expect("index out of range");
+        match svc.poll_ready() {
+            Ok(Async::Ready(())) => {
+                self.next_ready_index = Some(index);
+                Ok(Async::Ready(()))
+            }
+            Ok(Async::NotReady) => {
+                // became unready; so move it back there.
+                let (key, svc) = self
+                    .ready
+                    .swap_remove_index(index)
+                    .expect("invalid ready index");
+
+                // If a new version of this service has been added to the
+                // unready set, don't overwrite it.
+                if !self.pending_contains(&key) {
+                    self.push_service(key, svc);
+                }
+
+                Ok(Async::NotReady)
+            }
+            Err(e) => {
+                // failed, so drop it.
+                let (key, _) = self
+                    .ready
+                    .swap_remove_index(index)
+                    .expect("invalid ready index");
+                Err(error::Failed(key, e.into()))
+            }
+        }
+    }
+
+    /// Calls the next ready service.
+    ///
+    /// `poll_next_ready_index` must have returned Ready before this is invoked;
+    /// otherwise, this method panics.
+    pub fn call_next_ready(&mut self, req: Req) -> S::Future {
+        let index = self
+            .next_ready_index
+            .take()
+            .expect("poll_next_ready_index was not ready");
+
+        let (key, mut svc) = self
+            .ready
+            .swap_remove_index(index)
+            .expect("index out of range");
+
+        let fut = svc.call(req);
+
+        // If a new version of this service has been added to the
+        // unready set, don't overwrite it.
+        if !self.pending_contains(&key) {
+            self.push_service(key, svc);
+        }
+
+        fut
+    }
+}
+
+/// A default Service impleentation for ReadyCache`.
+///
+/// This service's `poll_ready` fails whenever an inner service fails and when
+/// there are no inner services in the cache. If this behavior is not desirable,
+/// users may call `poll_pending`, `poll_next_ready_index`, and `call_next_ready`
+/// directly.
+impl<K, S, Req> Service<Req> for ReadyCache<K, S, Req>
+where
+    K: Clone + Eq + Hash,
+    S: Service<Req>,
+    S::Error: Into<error::Error>,
+    error::Failed<K>: Into<error::Error>,
+{
+    type Response = S::Response;
+    type Error = error::Error;
+    type Future = future::MapErr<S::Future, fn(S::Error) -> Self::Error>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
```

Comment:
Should this document that it's not okay to call any other methods on `self` between getting `Ready` from this and calling `call`?

Reply:
I believe it's a general contract for service that poll_ready be called immediately before `call` (@carllerche ?), but I'll document something to that effect.